frontend/src/app/core/state/attachments/attachments.service.ts

```diff
@@ -0,0 +1,261 @@
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+//
+// See COPYRIGHT and LICENSE files for more details.
+//++
+
+import { Injectable } from '@angular/core';
+import {
+  HttpClient,
+  HttpHeaders,
+} from '@angular/common/http';
+import {
+  applyTransaction,
+  QueryEntity,
+} from '@datorama/akita';
+import {
+  from,
+  Observable,
+} from 'rxjs';
+import {
+  catchError,
+  map,
+  tap,
+} from 'rxjs/operators';
+import { AttachmentsStore } from 'core-app/core/state/attachments/attacments.store';
+import { IAttachment } from 'core-app/core/state/attachments/attachment.model';
+import { IHALCollection } from 'core-app/core/apiv3/types/hal-collection.type';
+import { ToastService } from 'core-app/shared/components/toaster/toast.service';
+import {
+  OpenProjectFileUploadService,
+  UploadFile,
+} from 'core-app/core/file-upload/op-file-upload.service';
+import { OpenProjectDirectFileUploadService } from 'core-app/core/file-upload/op-direct-file-upload.service';
+import { I18nService } from 'core-app/core/i18n/i18n.service';
+import { HalResource } from 'core-app/features/hal/resources/hal-resource';
+import { ApiV3Service } from 'core-app/core/apiv3/api-v3.service';
+import { HalLink } from 'core-app/features/hal/hal-link/hal-link';
+import isNewResource, { HAL_NEW_RESOURCE_ID } from 'core-app/features/hal/helpers/is-new-resource';
+import { ConfigurationService } from 'core-app/core/config/configuration.service';
+import { insertCollectionIntoState } from 'core-app/core/state/collection-store';
+
+@Injectable()
+export class AttachmentsResourceService {
+  protected store = new AttachmentsStore();
+
+  public query = new QueryEntity(this.store);
+
+  constructor(private readonly I18n:I18nService,
+    private readonly http:HttpClient,
+    private readonly apiV3Service:ApiV3Service,
+    private readonly fileUploadService:OpenProjectFileUploadService,
+    private readonly directFileUploadService:OpenProjectDirectFileUploadService,
+    private readonly configurationService:ConfigurationService,
+    private readonly toastService:ToastService) { }
+
+  /**
+   * Extension method that returns the same service instance after ensuring
+   * that a specific collection is fetched, if not available.
+   *
+   * @param key The collection key, of the required collection.
+   */
+  requireCollection(key:string):AttachmentsResourceService {
+    if (this.store.getValue().collections[key]) {
+      return this;
```

Comment:
The idea was to concat the service call like 

```
let x$ = this.attachmentsResourceService.requireCollection(key).query.select()
```

instead of 

```
if (keyNotLoadedInCollection) {
  this.attachentsResourceService.fetchAttachments(key).subscribe();
}

let x$ = this.attachmentsResourceService.query.select()
```

Sadly, in my only example, the collection can only be marked as required, if I am not on a new resource, so there is still the if.

So, you think this is no common pattern? I'm trying to think of a bit more sofisticated one. Join me please, if you have a great idea.

Reply:
I removed `return this` and changed return type to void.

I'd say, we need another round to come up with a fine pattern, of how the akita stores and services can implement an equivalent to `requireAndStream`